### PR TITLE
add and use more type traits in FEEvaluation

### PR DESCRIPTION
**Pull request by davydden**
these extension are need for vectors which are like sparse matrices. There should be no performance degradation for standard vectors, we just add a redirection mechanism in vector accessors.

### CODE REVIEW

**Comment by davydden** [comment]
there are failing tests 

````
The following tests FAILED:
	2134 - matrix_free/matrix_vector_faces_10.mpirun=5.debug (Failed)
	2136 - matrix_free/matrix_vector_faces_11.mpirun=5.debug (Failed)
	2142 - matrix_free/matrix_vector_faces_13.mpirun=5.debug (Failed)
	2144 - matrix_free/matrix_vector_faces_14.mpirun=2.debug (Failed)
	2146 - matrix_free/matrix_vector_faces_14.mpirun=3.debug (Failed)
	2150 - matrix_free/matrix_vector_faces_15.mpirun=5.debug (Failed)
	2156 - matrix_free/matrix_vector_faces_18.mpirun=4.debug (Failed)
	2158 - matrix_free/matrix_vector_faces_19.mpirun=5.debug (Failed)
	2160 - matrix_free/matrix_vector_faces_20.mpirun=3.debug (Failed)
	2162 - matrix_free/matrix_vector_faces_21.mpirun=3.debug (Failed)
	2164 - matrix_free/matrix_vector_faces_22.mpirun=3.debug (Failed)
	2166 - matrix_free/matrix_vector_faces_23.mpirun=3.debug (Failed)
```
that has to do with `const` reader
```
2166: An error occurred in line <1710> of file </Users/davydden/libs-sources/deal.ii/deal.ii/include/deal.II/lac/la_parallel_vector.h> in function
2166:     Number dealii::LinearAlgebra::distributed::Vector<double, dealii::MemorySpace::Host>::local_element(const dealii::LinearAlgebra::distributed::Vector::size_type) const
2166: The violated condition was:
2166:     local_index < local_size() || vector_is_ghosted == true
2166: Additional information:
2166:     You tried to read a ghost element of this vector, but it has not imported its ghost values.
```

It looks like a ghost DoF is being read without the vector being ghost (imported its ghost DoFs).

**Comment by kronbichler** [comment]
> Bottom line, I don't know much about `FEFaceEvaluation` but I would say this PR discovered a bug somewhere, I don't think it introduced one.

It's not a bug in the sense that results are incorrect, but rather a bug in the tracking of the `ghosted` state of the vector. The problematic line is this here:
https://github.com/dealii/dealii/blob/master/tests/matrix_free/matrix_vector_faces_common.h#L245-L246
which then redirects the control flow through some selective ghost update code which does not correctly set the 'vector_is_ghosted' flag. We should find a way to let the vector know about this state, which in turn requires that `struct internal::VectorDataExchange<int dim,typename Number>` is given the rights to do that. I do not like to make that internal class a friend, because it leaks a specific code to a basic linear algebra class, so we should probably rather include something like `LA::distributed::Vector::set_ghost_state(bool on_or_off)` or, instead, let the vector implement ghost exchange on a subset given by a second partitioner.

**Comment by davydden** [comment]
> we should probably rather include something like LA::distributed::Vector::set_ghost_state(bool on_or_off) 

I would go for this solution for the sake of this PR and document a (better?) alternative in an issues.

**Comment by davydden** [comment]
@kronbichler 

> The problematic line is this here:

Did you accidentally give a wrong link? I am trying to figure out where that call should be done...

**Comment by davydden** [comment]
I think I found it, must be `VectorDataExchange::update_ghost_values_finish`

**Comment by kronbichler** [comment]
The link I gave is the code that triggers that setting, but you are right that the code is in `update_ghost_values_finish` (and there should be an equivalent one in `reset_ghost_values`).

**Comment by davydden** [comment]
@kronbichler I fixed those test, the PR should be good now.

**Comment by davydden** [comment]
https://github.com/dealii/dealii/issues/7784 documents an alternative solution to the problem.

File: include/deal.II/matrix_free/fe_evaluation.h [hunk 1]
```diff
@@ -3580,17 +3724,29 @@ namespace internal
...
   }
 
-  // A class to use the same code to read from and write to vector
+
+
+  // Below, three classes (VectorReader, VectorSetter,
+  // VectorDistributorLocalToGlobal) implement the same interface and can be
+  // used to to read from vector, set elements of a vector and add to elements
+  // of the vector.
+  // FIXME: document interace
```
Comment (kronbichler) [inline include/deal.II/matrix_free/fe_evaluation.h]: I guess the current basic description is already better what we had before, so I suggest we skip the FIXME (we can always expand on this documentation - and as far as I remember, this is not part of the doxygen interface).

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: > so I suggest we skip the FIXME

ok, will do.

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: > as far as I remember, this is not part of the doxygen interface

true.

Comment (davydden) [reply include/deal.II/matrix_free/fe_evaluation.h]: @kronbichler fixed.

**Pull request merged**